File: packages/api-types/src/vibe.ts
import { z } from "zod";
import { apiResponse } from "./common.js";

// ─── POST /api/vibe/start ──────────────────────────────────────────────────

export const VibeStartRequestSchema = z.object({
  mode: z.enum(["mock", "llm"]).optional(),
});

export const VibeStartDataSchema = z.object({
  prompt: z.string(),
  step: z.literal("step1"),
});

export const VibeStartResponseSchema = apiResponse(VibeStartDataSchema);

export type VibeStartRequest = z.infer<typeof VibeStartRequestSchema>;
export type VibeStartData = z.infer<typeof VibeStartDataSchema>;
export type VibeStartResponse = z.infer<typeof VibeStartResponseSchema>;

// ─── POST /api/vibe/input ──────────────────────────────────────────────────

export const VibeInputRequestSchema = z.object({
  text: z.string(),
});

export const VibeInputDataSchema = z.union([
  z.object({ status: z.literal("input_received") }),
  z.object({ status: z.literal("no_more_questions") }),
  z.object({ next_question: z.string() }),
]);

export const VibeInputResponseSchema = apiResponse(VibeInputDataSchema);

export type VibeInputRequest = z.infer<typeof VibeInputRequestSchema>;
export type VibeInputData = z.infer<typeof VibeInputDataSchema>;
export type VibeInputResponse = z.infer<typeof VibeInputResponseSchema>;

// ─── POST /api/vibe/extract ────────────────────────────────────────────────

export const VibeExtractDataSchema = z.object({
  extracted: z.unknown(),
  status: z.string(),
  next: z.string().optional(),
  first_question: z.string().optional(),
  step2_error: z.string().optional(),
});

export const VibeExtractResponseSchema = apiResponse(VibeExtractDataSchema);

export type VibeExtractData = z.infer<typeof VibeExtractDataSchema>;
export type VibeExtractResponse = z.infer<typeof VibeExtractResponseSchema>;

File: lib/bropilot/api/handlers/vibe.ex
defmodule Bropilot.Api.Handlers.Vibe do
  @moduledoc """
  Handlers for Act 1 (Vibe Collection) endpoints.
  Manages an Act1.Worker process via the process registry.
  """

  import Bropilot.Api.Router, only: [json: 3]

  alias Bropilot.Pipeline.Act1.Worker

  @worker_name Bropilot.Api.VibeWorker

  def start(conn) do
    project_path = File.cwd!()
    recipe_dir = Path.join([project_path, ".bropilot", "recipe"])

    unless File.dir?(recipe_dir) do
      json(conn, 400, %{ok: false, error: "no .bropilot directory found — run `mix bro.init` first"})
    else
      # Stop any existing worker
      case Process.whereis(@worker_name) do
        nil -> :ok
        pid -> GenServer.stop(pid, :normal)
      end

      requested_mode = conn.body_params["mode"]

      {extraction_mode, llm_opts} =
        case requested_mode do
          "mock" ->
            {:mock, []}

          "llm" ->
            {:llm, [provider: Bropilot.LLM.provider()]}

          _ ->
            # Default: use LLM if configured, mock otherwise
            case Bropilot.LLM.provider() do
              :mock -> {:mock, []}
              provider -> {:llm, [provider: provider]}
            end
        end

      {:ok, pid} =
        Worker.start(
          project_path: project_path,
          recipe: recipe_dir,
          extraction_mode: extraction_mode,
          llm_opts: llm_opts
        )

      Process.register(pid, @worker_name)

      case Worker.run_step1(pid) do
        {:ok, prompt} ->
          json(conn, 200, %{ok: true, data: %{prompt: prompt, step: "step1"}})

        {:error, reason} ->
          json(conn, 500, %{ok: false, error: inspect(reason)})
      end
    end
  end

  def input(conn) do
    case Process.whereis(@worker_name) do
      nil ->
        json(conn, 400, %{ok: false, error: "vibe worker not started — call POST /api/vibe/start first"})

      pid ->
        text = conn.body_params["text"] || ""

        case Worker.submit_input(pid, text) do
          :ok ->
            # Check if we need to get the next question (step2)
            state = :sys.get_state(pid)

            if state.step == :step2 do
              case Worker.next_question(pid) do
                {:ok, :no_more_questions} ->
                  json(conn, 200, %{ok: true, data: %{status: "no_more_questions"}})

                {:ok, question} ->
                  json(conn, 200, %{ok: true, data: %{next_question: question}})
              end
            else
              json(conn, 200, %{ok: true, data: %{status: "input_received"}})
            end

          {:error, reason} ->
            json(conn, 500, %{ok: false, error: inspect(reason)})
        end
    end
  end

  def extract(conn) do
    case Process.whereis(@worker_name) do
      nil ->
        json(conn, 400, %{ok: false, error: "vibe worker not started — call POST /api/vibe/start first"})

      pid ->
        case Worker.extract(pid) do
          {:ok, data} ->
            state = :sys.get_state(pid)

            response =
              case state.step do
                :step1_done ->
                  %{extracted: data, status: "step1_done", next: "step2"}

                :complete ->
                  %{extracted: data, status: "complete"}

                other ->
                  %{extracted: data, status: to_string(other)}
              end

            # If step1 is done, auto-start step2
            if state.step == :step1_done do
              case Worker.run_step2(pid) do
                {:ok, first_question} ->
                  json(conn, 200, %{
                    ok: true,
                    data: Map.put(response, :first_question, first_question)
                  })

                {:error, reason} ->
                  json(conn, 200, %{
                    ok: true,
                    data: Map.put(response, :step2_error, inspect(reason))
                  })
              end
            else
              json(conn, 200, %{ok: true, data: response})
            end

          {:error, reason} ->
            json(conn, 400, %{ok: false, error: inspect(reason)})
        end
    end
  end
end
